src/endpoints/patch-session.js

```diff
@@ -14,29 +15,160 @@ module.exports = cors(async (req, res) => {
+        const content = JSON.parse(sample.content)
+        const annotation = JSON.parse(sample.annotation)
+        samples.push(Object.assign({}, content, {annotation}))
+    })
+
+    newJSON.samples = samples
+
     return send(res, 200, {latestVersion, hashOfLatestState: hash(newJSON)})
+})
+
+const patchCreateSamples = (async (sessionId, samplesToAdd) => {
+    const samplesQueries = []
+    if (samplesToAdd.length) {
+        let sessionSampleId = -1
+        const latestSample = db.prepare('SELECT * FROM sample_state WHERE session_short_id = ? ORDER BY session_sample_id DESC LIMIT 1').get(sessionId);
+        if (latestSample) {
+            sessionSampleId = parseInt(latestSample.session_sample_id)
+        }
+
+        samplesToAdd.forEach((sample) => {
+            sessionSampleId += 1
+            samplesQueries.push(db.prepare('INSERT INTO sample_state (session_short_id, session_sample_id, content) VALUES (?, ?, ?)').run(sessionId, sessionSampleId, JSON.stringify(sample)));
+        })
+    }
+    return samplesQueries
+})
+
+const patchRemoveSamples = (async (sessionId, patches) => {
+    if (patches.length) {
+        await asyncForEach(patches, async (patch) => {
+            const pathArray = patch.path.split('/')
+            const indexToRemove = pathArray[2]
+
+            await db.prepare('DELETE FROM sample_state WHERE session_short_id = ? AND session_sample_id = ? ').run(sessionId, indexToRemove);
+            await db.prepare(`
+                INSERT INTO sample_state 
+                    (session_short_id, session_sample_id, version, content, summary, annotation) 
+                SELECT 
+                    session_short_id, session_sample_id - 1, version + 1, content, summary, annotation
+                FROM 
+                    latest_sample_state
+                WHERE 
+                    session_short_id = ? AND session_sample_id > ?`).run(sessionId, indexToRemove);
+
+            await db.prepare(`
+                DELETE FROM sample_state 
+                WHERE sample_state_id IN (
+                    SELECT sample_state_id
+                    FROM sample_state
+                    WHERE session_sample_id = (
+                        SELECT session_sample_id
+                        FROM latest_sample_state 
+                        WHERE session_short_id = ? 
+                        ORDER BY session_sample_id DESC 
+                        limit 1
+                    )
+                )`).run(sessionId);
+        })
+    }
+})
+
+const patchSamplesAnnotation = (async (sessionId, samplePatches) => {
+
+    const sampleIds = {}
+    samplePatches.forEach(patch => {
+        const pathArray = patch.path.split('/')
+        const sessionSampleId = pathArray[2]
+        sampleIds[sessionSampleId] = {sessionSampleId}
+    })
+
+    const samplesArray = []
+    const samples = db.prepare('SELECT * FROM latest_sample_state WHERE session_short_id = ? ORDER BY session_sample_id ASC').all(sessionId);
```

Comment:
don't load all the samples into memory, there's no need to (just apply each patch to the the relevant sample)

Reply:
I need to get the same indexes so applyPatch would find and update the requested index